File: apps/trajectories.cc
#include <cassert>
#include <fstream>
#include <iostream>

#include <gflags/gflags.h>

#include <boolean_network/dynamics.h>
#include <boolean_network/network.h>
#include <boolean_network/writer.h>

DEFINE_int32(N, 0, "N");
DEFINE_int32(k, 0, "k");
DEFINE_double(a, 0, "a");
DEFINE_int32(num_trials, 1, "num-trials");
DEFINE_int32(num_steps, 0, "num-steps");
DEFINE_uint32(seed, 0, "seed");
DEFINE_string(tag, "", "tag");

int main(int argc, char** argv) {
  gflags::SetUsageMessage(
    "Record the states of a boolean network with randomized initial starting states."
  );
  gflags::ParseCommandLineFlags(&argc, &argv, true);

  assert(FLAGS_N > 0);
  assert(FLAGS_k > 0);
  assert(FLAGS_k <= FLAGS_N);
  assert(FLAGS_num_trials > 0);
  assert(FLAGS_num_steps > 0);
  assert(FLAGS_a > 0);

  network::NetworkConfig network_config;
  network_config.N = FLAGS_N;
  network_config.k = FLAGS_k;
  network_config.a = FLAGS_a;
  network_config.seed = FLAGS_seed;

  network::BooleanThresholdNetwork network(network_config);
  GetNetwork(network_config, &network);

  network::DynamicsConfig dynamics_config;
  dynamics_config.num_trials = FLAGS_num_trials;
  dynamics_config.num_steps = FLAGS_num_steps;

  network::MetaData metadata;
  metadata.tag = FLAGS_tag;
  metadata.start_time = std::chrono::system_clock::now();

  network::Trajectories trajectories;
  network::ComputeTrajectories(network, dynamics_config, &trajectories);

  std::cout << "done" << std::endl;
  metadata.end_time = std::chrono::system_clock::now();

  const std::string output_file_name = network::GetOutputFileName(
    "trajectories-",
    metadata.start_time
  );
  std::ofstream ofs{"data/" + output_file_name + ".json"};

  network::WriteTrajectoriesToStream(
    trajectories,
    network_config,
    dynamics_config,
    metadata,
    &ofs
  );
}

File: src/bin/trajectories.rs
use clap::Parser;
use chrono::Utc;

// Bring in your library’s types & functions:
use boolean_threshold_network::{
  types::NetworkConfig,
  types::DynamicsConfig,
  types::BooleanThresholdNetwork,
  types::MetaData,
  dynamics::compute_trajectories,
  writer::write_protobuf,
  // GetOutputFileName,
  // write_trajectories_to_stream,
};

// use generated::boolean_threshold_network::;
pub mod pb {
  include!(concat!(env!("OUT_DIR"), "/boolean_threshold_network.rs"));
}

/// Record the states of a boolean network with randomized initial starting states.
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
  /// Number of nodes in the network
  #[arg(short = 'N', long)]
  N: usize,

  /// Threshold parameter
  #[arg(long)]
  gamma: f64,

  /// How many independent trials to run
  #[arg(long, default_value_t = 1)]
  num_trials: usize,

  /// How many steps per trial
  #[arg(long)]
  num_steps: usize,

  /// RNG seed for the network itself
  #[arg(long, default_value_t = 0)]
  network_seed: usize,

  /// RNG seed for the dynamics
  #[arg(long, default_value_t = 0)]
  dynamics_seed: usize,

  /// A tag to stick in the metadata
  #[arg(long, default_value = "")]
  tag: String,
}


fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
  let args = Args::parse();

  assert!(args.N > 0, "N must be > 0");
  assert!(args.num_trials > 0, "num_trials must be > 0");
  assert!(args.num_steps > 0, "num_steps must be > 0");

  let network_config = NetworkConfig {
    N: args.N,
    gamma: args.gamma,
    seed: args.network_seed,
  };
  let dynamics_config = DynamicsConfig {
    num_trials: args.num_trials,
    num_steps: args.num_steps,
    seed: args.dynamics_seed,
  };

  let mut network = BooleanThresholdNetwork::new(&network_config);

  let mut metadata = MetaData {
    tag: args.tag.clone(),
    start_time: Utc::now(),
    end_time: Utc::now(),
  };

  let trajectories = compute_trajectories(&mut network, &dynamics_config);
  metadata.end_time = Utc::now();

  write_protobuf(
    "data/out.trajpb",
    &trajectories,
    &metadata,
    &dynamics_config,
    &network_config,
    &network,
  )?;
  println!("Wrote {}", "out.trajpb");
  println!("done");

  // let fname = GetOutputFileName("trajectories-", metadata.start_time);
  // let path = format!("data/{}.json", fname);

  // let file = File::create(&path)?;
  // let mut writer = BufWriter::new(file);
  // write_trajectories_to_stream(
  //   &trajectories,
  //   &network,
  //   &network_config,
  //   &dynamics_config,
  //   &metadata,
  //   &mut writer,
  // )?;

  Ok(())
}
